src/buffer_test.cc

```diff
@@ -124,4 +124,152 @@ TEST_F(BufferTest, SizeMatrixPaddedStd430) {
   EXPECT_EQ(12U * sizeof(int32_t), b.GetSizeInBytes());
 }
 
+TEST_F(BufferTest, GetHistogramForChannelGradient) {
+  TypeParser parser;
+  auto type = parser.Parse("R8G8B8A8_UINT");
+  Format fmt(type.get());
+
+  std::vector<Value> values(40);
+  for (uint64_t i = 0; i < values.size(); i += 4)
+    values[i + 2].SetIntValue(i / 4 * 25);
+
+  Buffer b(BufferType::kColor);
+  b.SetFormat(&fmt);
+  b.SetData(values);
+
+  std::vector<uint64_t> bins = b.GetHistogramForChannel(2, 256);
+  for (uint64_t i = 0; i < values.size(); i += 4)
+    EXPECT_EQ(1, bins[i / 4 * 25]);
+}
+
+TEST_F(BufferTest, GetHistogramForChannelAllBlack) {
+  TypeParser parser;
+  auto type = parser.Parse("R8G8B8A8_UINT");
+  Format fmt(type.get());
+
+  std::vector<Value> values(40);
+  for (uint64_t i = 0; i < values.size(); i++)
+    values[i].SetIntValue(0);
+
+  Buffer b(BufferType::kColor);
+  b.SetFormat(&fmt);
+  b.SetData(values);
+
+  for (uint8_t i = 0; i < 4; i++) {
+    std::vector<uint64_t> bins = b.GetHistogramForChannel(i, 256);
+    for (uint64_t y = 0; y < values.size(); y++)
+      EXPECT_EQ(10, bins[0]);
+  }
+}
+
+TEST_F(BufferTest, GetHistogramForChannelAllWhite) {
+  TypeParser parser;
+  auto type = parser.Parse("R8G8B8A8_UINT");
+  Format fmt(type.get());
+
+  std::vector<Value> values(40);
+  for (uint64_t i = 0; i < values.size(); i++)
+    values[i].SetIntValue(std::numeric_limits<uint8_t>::max());
+
+  Buffer b(BufferType::kColor);
+  b.SetFormat(&fmt);
+  b.SetData(values);
+
+  for (uint8_t i = 0; i < 4; i++) {
+    std::vector<uint64_t> bins = b.GetHistogramForChannel(i, 256);
+    for (uint64_t y = 0; y < values.size(); y++)
+      EXPECT_EQ(10, bins[255]);
+  }
+}
+
+TEST_F(BufferTest, CompareHistogramEMDToleranceFalse) {
+  TypeParser parser;
```

Comment:
Please add something like this to the top of the function:

```cpp
// Creates two sets of equal pixel values, except for one pixel that has +50 in its red channel.
// Compares the histograms to see if they are equal with a low threshold, which we expect to fail.
```

Reply:
Done.